kotest-assertions/kotest-assertions-core/src/jvmTest/kotlin/com/sksamuel/kotest/inspectors/FilterMatchingTest.kt

```
@@ -0,0 +1,57 @@
+package com.sksamuel.kotest.inspectors
+
+import io.kotest.assertions.assertSoftly
+import io.kotest.core.spec.style.FunSpec
+import io.kotest.inspectors.filterMatching
+import io.kotest.matchers.collections.shouldContainExactly
+import io.kotest.matchers.ints.shouldBeEven
+import io.kotest.matchers.longs.shouldBeLessThan
+import io.kotest.matchers.sequences.shouldContainExactly
+import io.kotest.matchers.shouldBe
+import io.kotest.matchers.string.shouldHaveLength
+import io.kotest.matchers.string.shouldNotContain
+import io.kotest.matchers.types.shouldBeInstanceOf
+import kotlin.system.measureTimeMillis
+import kotlin.time.Duration.Companion.seconds
+
+class FilterMatchingTest : FunSpec({
+   test("Filtering empty list should return empty list") {
+      emptyList<String>().filterMatching {
+         it shouldHaveLength 5
+      } shouldBe emptyList()
+   }
+
+   test("Filtering should only return matching elements") {
+      arrayOf("a", "bb", "ccc", "dddd").filterMatching {
+         it shouldHaveLength 1
+      } shouldBe listOf("a")
+   }
+
+   test("Filtering applies all assertions to each element") {
+      arrayOf("a", "b", "bb", "ccc", "dddd").filterMatching {
+         it shouldHaveLength 1
+         it shouldNotContain "b"
+      } shouldBe listOf("a")
+   }
+
+   test("Filtering should work in assertSoftly context") {
+      assertSoftly {
+         1 shouldBe 1
+         arrayOf(1, 2, 3).filterMatching { it.shouldBeEven() } shouldBe listOf(2)
+         listOf(1, 2, 3).filterMatching { it.shouldBeEven() } shouldBe listOf(2)
+
+         sequenceOf(1, 2, 3)
+            .filterMatching { it.shouldBeEven() }
+            .shouldBeInstanceOf<Sequence<Int>>()
+            .shouldContainExactly(2)
+      }
+   }
+
+   test("Filtering sequences should be done lazily").config(invocationTimeout = 1.seconds) {
+      // Generating an infinite sequence. If the filtering is not done lazily, this test will timeout
+      generateSequence(1) { it + 1 }
```

will this test fail faster if the sequence yields 100 elements and then throws an exception?

---

It currently runs in (rounded to) 0ms on my machine, so I would assume so.

---

it succeeds right away, but should it ever fail, it would take a whole second - care to speed that up?

---

I've had issues in the past with GH actions runners which sometimes pause threads for long enough to cause sub-second timeouts to become flaky. 

But since this code actually never suspends / yields, there would never be a point in time where the `InvocationTimeoutInterceptor`'s `withTimeout` gets to run, so it wouldn't be a problem. But that also means that the test is broken as-is. Will adjust it a bit, but I think we should keep the timeout at 1s to prevent flakiness from creeping in.

---

> I think we should keep the timeout at 1s to prevent flakiness from creeping in.

I'd suggest designing away the very need for timeout: replace `generateSequence(1) { it + 1 }`
with something like
```kotlin
sequence {
(1..100).forEach { yield(it) }
throw Exception("Oops!")
}
```
then we won't need any timeout at all

---

Clever 😃 sounds good, should cover the scenario better.